[quant][fx][graphmode] Add support for sigmoid

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46006 [wip][quant][graphmode][fx] Extend custom module to support multiple outputs
* #45920 [quant][fx][graphmode][api] Change API for custom module
* #45919 [quant][fx][graphmode] Run symbolic_trace in quantization
* **#45885 [quant][fx][graphmode] Add support for sigmoid**
* #45884 [quant][eagermode] Add support for sigmoid module

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24129114](https://our.internmc.facebook.com/intern/diff/D24129114)

jerryzh168 committed:
[quant][fx][graphmode] Add support for sigmoid

Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add support for sigmoid"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24129114](https://our.internmc.facebook.com/intern/diff/D24129114)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add support for sigmoid"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24129114](https://our.internmc.facebook.com/intern/diff/D24129114)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add support for sigmoid"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24129114](https://our.internmc.facebook.com/intern/diff/D24129114)

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1120,46 +1120,86 @@ def forward(self, x):
-        self._test_activation_impl(nn.Hardswish, F.hardswish, nnq.Hardswish, torch.ops.quantized.hardswish)
+        for inplace in [True, False]:
+            extra_kwargs = {'inplace': inplace}
+            self._test_activation_impl(
```
Can we pass `inplace` as an argument to `_test_activation_impl` to avoid this extra look in the test?

jerryzh168 replied:
not sure I understand, I think current way of passing argument is more general than just passing an inplace argument

supriyar replied:
Oh, my suggestion was to move the loop over `inplace` inside `_test_activation_impl`.

jerryzh168 replied:
oh, we can't do that since not all activation functions has `inplace` argument

supriyar reviewed (approved):
LG, left a small comment.

raghuramank100 commented on torch/quantization/quantization_mappings.py:
```diff
@@ -86,6 +86,9 @@
 FLOAT_TO_QUANTIZED_OPERATOR_MAPPINGS = {
     F.elu: torch._ops.ops.quantized.elu,
     F.leaky_relu: torch._ops.ops.quantized.leaky_relu,
+    F.sigmoid: torch._ops.ops.quantized.sigmoid,
```
Why is this specific to sigmoid? Do we need three forms for every functional?

jerryzh168 replied:
not necessarily, `x.leaky_relu()` is not supported I think: https://pytorch.org/docs/stable/search.html?q=leaky_relu&check_keywords=yes&area=default#

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -400,14 +401,18 @@ def convert(self, quantizer, node, load_arg, debug=False):
             return NotImplemented
-        assert node.op in ['call_module', 'call_function'], 'Only call_module and ' + \
-            'call_function are handled in DefaultNode'
+        assert node.op in ['call_module', 'call_function', 'call_method'], 'Only call_module and ' + \
```
Should the text of the assert change to reflect that call methods are supported?

jerryzh168 replied:
sure

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add support for sigmoid"


Summary:

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_sigmoid

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24129114](https://our.internmc.facebook.com/intern/diff/D24129114)

[ghstack-poisoned]

jerryzh168 commented:
we'll use fixed quantization parameter for sigmoid since there are implementations that only implements sigmoid with fixed quantization parameters, e.g. qnnpack: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qnnpack/src/sigmoid.c#L72, and it uses LUT as an optimization. this can't be done if we support arbitrary output_scale and output_zero_point.
